Document features (and add fancy gifs) (#9)
tarides/ocaml-eglot: README.md

user experience as close as possible to that offered by the Emacs mode
[Merlin](https://ocaml.github.io/merlin/editor/emacs/).

## Features

### Browsing errors

Eglot relies on
[Flymake](https://www.gnu.org/software/emacs/manual/html_node/emacs/Flymake.html)
for error diagnosis. OCaml-eglot offers two functions for quickly
navigating through errors:

- `ocaml-eglot-error-next` (<kbd>C-c</kbd> <kbd>C-x</kbd>): jump to
  the next error
- `ocaml-eglot-error-prev` (<kbd>C-c</kbd> <kbd>C-c</kbd>): jump to
  the previous error

![Error navigation example](media/error-navigation.gif)

### Jump to definition

Eglot relies on
[Xref](https://www.gnu.org/software/emacs/manual/html_node/emacs/Xref.html)
to index cross-references. OCaml-eglot provides a shortcut to quickly
jump to the definition:

- `ocaml-eglot-locate` (<kbd>C-c</kbd> <kbd>C-l</kbd>): jump to
  definition

![Jump to definition example](media/locate.gif)

### Infer Interface

Used to infer the type of an interface file. If the buffer is not
empty, a prompt will ask for confirmation to overwrite the buffer

> contents:

Reviewer: Reading only the text, it's unclear of which "non-empty buffer" your talking about.

I guess the command should be called from within an implementation file and will create a new buffer for the interface. It is that new buffer that might erase an existing one right ? I think it could be made a bit clearer.

Reviewer: And now that I looked at the gif, I understand it is the other way around :-)

For me it makes more sense to (also) be able to call infer-interface from the implementation itself but I don't have a strong opinion.

> - `ocaml-eglot-infer-interface`: infer the current interface file

Reviewer: Isn't it rather:

```suggestion
- `ocaml-eglot-infer-interface`: infer the interface for the current implementation file
```

![Infer Interface example](media/infer-interface.gif)

### Find Alternate file

OCaml-eglot allows you to quickly switch from the implementation file
to the interface file and _vice versa_. If the interface file does not
exist, a prompt can be used to generate it (using type inference,
based on `ocaml-eglot-infer-inteface`):

- `ocaml-eglot-alternate-file` (<kbd>C-c</kbd> <kbd>C-a</kbd>): switch
  from the implementation file to the interface file and _vice versa_

![Find Alternate File example](media/alternate-file.gif)

### Get Documentation

> Although the `Hover` primitive in the LSP protocol can be used to
> conveiniently display value documentation, it is possible to invoke it
> manually:

Reviewer: ```suggestion
Although the `Hover` primitive in the LSP protocol can be used to
conveniently display value documentation, it is also possible to query for it
specifically:
```

- `ocaml-eglot-document` (<kbd>C-c</kbd> <kbd>C-d</kbd>): documents
  the expression below the cursor.
- `ocaml-eglot-document-identifier`: enables you to enter an
  identifier (present in the environment) and return its
  documentation.

![Get Documentation Example](media/document.gif)

### Construct Expression

> Enables you to navigate between the different types of holes in a
> document and interactively subsitute them:

Reviewer: ```suggestion
Enables you to navigate between the different typed-holes (`_`) in a
document and interactively substitute them:
```

- `ocaml-eglot-hole-next`: jump to the next hole
- `ocaml-eglot-hole-prev`: jump to the previous hole

> - `ocaml-eglot-construct`: Opens up a list of valid substitutions to

Reviewer: ```suggestion
- `ocaml-eglot-construct`: open up a list of valid substitutions to
```

  fill the hole

![Construct Example](media/construct.gif)

If the `ocaml-eglot-construct` command is prefixed by an argument, ie:
`C-u M-x ocaml-eglot-construct`, the command will also search for
valid candidates in the current environment:

![Construct with prefix-arg Example](media/construct2.gif)

### Source Browsing

OCaml-eglot allows you to navigate semantically in a buffer, passing
from an expression to the parent `let`, the parent `module`, the
parent `fun` and the parent `match` expression. It is also possible to
navigate between pattern matching cases:

> 

Reviewer: We should indicate the targets for pattern matching cases. Even better, could they appear as suggestions when calling the command ?

The custom request can also return all possible targets with the corresponding jump, that might be a better workflow ?
- Execute the command which wait for the CR answer
- Ask the user which of the available jumps she wants to perform.

Author: Oh yes, non giving a target will compute every available targets?

Reviewer: Yes !
